decimaldate.js

```diff
@@ -9,129 +9,157 @@ decimaldate.DECIMALPLACES = 6;
-    if (decdate >= 0) {
-        yearint = parseInt(Math.abs(Math.floor(decdate)));
-        plusminus = '';
+    // remove the artificial +1 that we add to make positive dates look intuitive
+    const truedecdate = decdate - 1;
+    const ispositive = truedecdate > 0;
+
+    // get the integer year
+    if (ispositive) {
+        yearint = Math.floor(truedecdate) + 1;
     }
     else {
-        yearint = parseInt(Math.abs(Math.ceil(decdate)));
-        plusminus = '-';
+        yearint = -Math.abs(Math.floor(truedecdate));  // ISO 8601 shift and year<=0 by 1, 0=1BCE, -1=2BCE
     }
 
-    const yearstring = yearint + "";
-    const dty = decimaldate.daysinyear(yearstring);
-    let targetday = dty * (Math.abs(decdate) % 1);
-    targetday = decdate >= 0 ? Math.ceil(targetday) : Math.floor(targetday);
-    if (decdate < 0) targetday = dty - targetday;
+    // how many days in year X decimal portion = number of days into the year
+    // if it's <0 then we count backward from the end of the year, instead of forward into the year
+    const dty = decimaldate.daysinyear(yearint);
+    let targetday = dty * (Math.abs(truedecdate) % 1);
+    if (ispositive) targetday = Math.ceil(targetday);
+    else targetday = dty - Math.floor(targetday);
 
     // count up days months at a time, until we reach our target month
-    // the the remainder days is the day of the month, offset by 1 cuz we count from 0
-    const months = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'];
-    let monthstring;
+    // then the remainder (days) is the day of that month
+    let monthint;
     let dayspassed = 0;
-    for (var i=0, l=months.length; i<l; i++) {
-        monthstring = months[i];
-
-        const dtm = decimaldate.daysinmonth(yearstring, monthstring);
+    for (let m = 1; m <= 12; m++) {
+        monthint = m;
+        const dtm = decimaldate.daysinmonth(yearint, monthint);
         if (dayspassed + dtm < targetday) {
             dayspassed += dtm;
         }
         else {
             break;
         }
     }
-
-    const daynumber = targetday - dayspassed;
-    const daystring = (daynumber < 10 ? "0" : "") + daynumber;
-
-    return `${plusminus}${yearstring}-${monthstring}-${daystring}`;
+    const dayint = targetday - dayspassed;
+
+    // make string output
+    // months and day as 2 digits
+    // ISO 8601 shift year<=0 by 1, 0=1BCE, -1=2BCE
+    const monthstring = monthint.toString().padStart(2, '0');
+    const daystring = dayint.toString().padStart(2, '0');
+    let yearstring;
+    if (yearint > 0) yearstring = yearint.toString().padStart(4, '0');  // just the year as 4 digits
+    else if (yearint == -1) yearstring = (Math.abs(yearint + 1).toString().padStart(4, '0'));  // BCE offset by 1 but do not add a - sign
+    else yearstring = '-' + (Math.abs(yearint + 1).toString().padStart(4, '0'));  // BCE offset by 1 and add  - sign
+
+    return `${yearstring}-${monthstring}-${daystring}`;
```

Comment:
It would be simpler to create a `Date` object and call [`toISOString`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString). We don’t care about the time, so you can trim it at the `T`.

Note that, per ISO&nbsp;8601, `toISOString` returns negative years that are six digits long, but we’re violating that part of the specification consistently throughout the stack. (Don’t worry about it just yet…) For now, we can correct for that situation with something like `replace(/^-00(\d{4})/, "-$1")`.